### PR TITLE
SQL based notes panel

### PR DESCRIPTION
**Pull request by quardbreak**

### CODE REVIEW

File: code/modules/admin/sql_notes.dm [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	if(!dbcon.IsConnected())
+		to_chat(usr, "<span class='danger'>Failed to establish database connection.</span>")
+		return
+
+	if(!target_ckey)
+		var/new_ckey = ckey(input(usr,"Who would you like to add a note for?","Enter a ckey",null) as text)
+		if(!new_ckey)
+			return
+		new_ckey = sanitizeSQL(new_ckey)
+		var/DBQuery/query_find_ckey = dbcon.NewQuery("SELECT ckey FROM erro_player WHERE ckey = '[new_ckey]'")
```
Comment (Andrew-Fall) [inline code/modules/admin/sql_notes.dm]: Needs backticks for columns and tables
dbcon.NewQuery("SELECT \`ckey\` FROM \`erro_player\` WHERE \`ckey\` = '[new_ckey]'")

File: code/modules/admin/sql_notes.dm [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	if(!adminckey)
+		adminckey = usr.ckey
+		if(!adminckey)
+			return
+	var/admin_sql_ckey = sanitizeSQL(adminckey)
+	if(!server)
+		if (config && config.server_name)
+			server = config.server_name
+	server = sanitizeSQL(server)
+	var/DBQuery/query_noteadd = dbcon.NewQuery("INSERT INTO erro_messages (type, targetckey, adminckey, text, timestamp, server) VALUES ('note', '[target_sql_ckey]', '[admin_sql_ckey]', '[notetext]', '[timestamp]', '[server]')")
```
Comment (Andrew-Fall) [inline code/modules/admin/sql_notes.dm]: dbcon.NewQuery("INSERT INTO \`erro_messages\` (\`type\`, \`targetckey\`, \`adminckey\`, \`text\`, \`timestamp\`, \`server\`) VALUES ('note', '[target_sql_ckey]', '[admin_sql_ckey]', '[notetext]', '[timestamp]', '[server]')")

File: code/modules/admin/sql_notes.dm [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+	if(!check_rights(R_INVESTIGATE, 1, usr)) return
+
+	if(!dbcon.IsConnected())
+		to_chat(usr, "<span class='danger'>Failed to establish database connection.</span>")
+		return
+	if(!note_id)
+		return
+	note_id = text2num(note_id)
+	var/DBQuery/query_find_note_del = dbcon.NewQuery("SELECT targetckey, text, adminckey FROM erro_messages WHERE id = [note_id] and type = 'note'")
```
Comment (Andrew-Fall) [inline code/modules/admin/sql_notes.dm]: dbcon.NewQuery("SELECT \`targetckey\`, \`text\`, \`adminckey\` FROM \`erro_messages\` WHERE \`id\` = [note_id] and \`type\` = 'note'")

File: code/modules/admin/sql_notes.dm [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	if(!query_find_note_del.Execute())
+		var/err = query_find_note_del.ErrorMsg()
+		log_game("SQL ERROR obtaining targetckey, text, adminckey from erro_messages table. Error : \[[err]\]\n")
+		return
+	if(query_find_note_del.NextRow())
+		ckey = query_find_note_del.item[1]
+		notetext = query_find_note_del.item[2]
+		adminckey = query_find_note_del.item[3]
+	var/DBQuery/query_del_note = dbcon.NewQuery("DELETE FROM erro_messages WHERE id = [note_id]")
```
Comment (Andrew-Fall) [inline code/modules/admin/sql_notes.dm]: dbcon.NewQuery("DELETE FROM \`erro_messages\` WHERE \`id\` = [note_id]")

File: code/modules/admin/sql_notes.dm [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+	if(!dbcon.IsConnected())
+		to_chat(usr, "<span class='danger'>Failed to establish database connection.</span>")
+		return
+	if(!note_id)
+		return
+	note_id = text2num(note_id)
+	var/target_ckey
+	var/sql_ckey = sanitizeSQL(usr.ckey)
+	var/DBQuery/query_find_note_edit = dbcon.NewQuery("SELECT targetckey, text, adminckey FROM erro_messages WHERE id = [note_id] and type = 'note'")
```
Comment (Andrew-Fall) [inline code/modules/admin/sql_notes.dm]: dbcon.NewQuery("SELECT \`targetckey\`, \`text\`, \`adminckey\` FROM \`erro_messages\` WHERE \`id\` = [note_id] and \`type\` = 'note'")

File: code/modules/admin/sql_notes.dm [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+		target_ckey = query_find_note_edit.item[1]
+		var/old_note = query_find_note_edit.item[2]
+		var/adminckey = query_find_note_edit.item[3]
+		var/new_note = input("Input new note", "New Note", "[old_note]") as message
+		if(!new_note)
+			return
+		new_note = sanitizeSQL(new_note)
+		var/edit_text = "Edited by [sql_ckey] on [SQLtime()] from<br>[old_note]<br>to<br>[new_note]<hr>"
+		edit_text = sanitizeSQL(edit_text)
+		var/DBQuery/query_update_note = dbcon.NewQuery("UPDATE erro_messages SET text = '[new_note]', lasteditor = '[sql_ckey]', edits = CONCAT(IFNULL(edits,''),'[edit_text]') WHERE id = [note_id]")
```
Comment (Andrew-Fall) [inline code/modules/admin/sql_notes.dm]: Backticks

File: code/modules/admin/sql_notes.dm [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+		navbar += "|<a href='?_src_=holder;shownote=[letter]'>\[[letter]\]</a>"
+	navbar += "<br><form method='GET' name='search' action='?'>\
+	<input type='hidden' name='_src_' value='holder'>\
+	<input type='text' name='notessearch' value='[index]'>\
+	<input type='submit' value='Search'></form>"
+	if(!linkless && check_rights(R_INVESTIGATE, 0, usr))
+		output = navbar
+	if(target_ckey)
+		var/target_sql_ckey = sanitizeSQL(target_ckey)
+		var/DBQuery/query_get_notes = dbcon.NewQuery("SELECT id, timestamp, text, adminckey, lasteditor, server FROM erro_messages WHERE targetckey = '[target_sql_ckey]' and type = 'note' ORDER BY timestamp")
```
Comment (Andrew-Fall) [inline code/modules/admin/sql_notes.dm]: Backticks

File: code/modules/admin/sql_notes.dm [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+		if(!isnum(index))
+			index = sanitizeSQL(index)
+		switch(index)
+			if(1)
+				search = "^."
+			if(2)
+				search = "^\[^\[:alpha:\]\]"
+			else
+				search = "^[index]"
+		var/DBQuery/query_list_notes = dbcon.NewQuery("SELECT DISTINCT targetckey FROM erro_messages WHERE targetckey REGEXP '[search]' ORDER BY targetckey")
```
Comment (Andrew-Fall) [inline code/modules/admin/sql_notes.dm]: Backticks

File: code/modules/admin/topic.dm [hunk 9]
```diff
@@ -1675,28 +1675,69 @@
...
+		var/target_ckey = href_list["shownoteckey"]
+		show_note(target_ckey)
+
+	if(href_list["notessearch"])
+		var/target = href_list["notessearch"]
+		show_note(index = target)
+
+	if(href_list["noteedits"])
+		var/note_id = sanitizeSQL("[href_list["noteedits"]]")
+		var/DBQuery/query_noteedits = dbcon.NewQuery("SELECT edits FROM erro_messages WHERE id = '[note_id]'")
```
Comment (Andrew-Fall) [inline code/modules/admin/topic.dm]: Backticks

**Review by Andrew-Fall — CHANGES_REQUESTED** [review]
SQL needs backticks

**Review by Andrew-Fall — DISMISSED** [review]

**Comment by quardbreak** [comment]
Require help with testing. Unable to test locally.

**Comment by quardbreak** [comment]
Reconverting from draft for review. DNM till we don't test this.

File: code/modules/admin/sql_notes.dm [hunk 10]
```diff
@@ -0,0 +1,186 @@
+/proc/add_note(target_ckey, notetext, timestamp, adminckey, logged = 1, server)
+	if(!check_rights(R_INVESTIGATE, 1, usr)) return
+
+	if(!dbcon.IsConnected())
```
Comment (MistakeNot4892) [inline code/modules/admin/sql_notes.dm:5]: `!dbcon?.IsConnected`

File: code/modules/admin/sql_notes.dm [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+		to_chat(usr, "<span class='danger'>Failed to establish database connection.</span>")
```
Comment (MistakeNot4892) [inline code/modules/admin/sql_notes.dm]: `SPAN_WARNING`

File: code/modules/admin/sql_notes.dm (hunk 10 above)
Comment (MistakeNot4892) [reply code/modules/admin/sql_notes.dm:5]: nvm

**Comment by MistakeNot4892** [comment]
I keep forgetting this exists  but it's now two weeks stale and last update was that it needed testing with a MySQL server before being mergable, so feel free to reopen when it has been tested. I don't have any experience using MySQL and no access to an instance so can't really offer to test it myself. Could possibly ask Lohi nicely?

**Pull request closed without merging**